### ISSUE TITLE
[rl] PPG aux-phase distillation KL is exactly zero across all minibatches

### ISSUE THREAD

**Issue by anthonytorlucci** [anthonytorlucci/rlevo#319]
Found incidentally while migrating PPG for #167. **Pre-existing at HEAD — not a regression from that change.**

## Observation

While A/B-diffing seeded CartPole runs to prove #167's reordering was bit-identical, the baseline and patched runs agreed exactly — including on this: PPG's auxiliary-phase distillation KL is bit-exactly `0x00000000` across **all 24 minibatches** (6 epochs × 24), on both pristine HEAD and the migrated tree.

## Why it looks wrong

The aux phase distills the policy against a snapshot taken *before* the policy phase stepped. Once the policy has taken optimizer steps, its output distribution should have moved off the snapshot, so the distillation KL should be a small **non-zero** value. An exact `0.0` — not `1e-8`, but every bit zero, every minibatch — is the signature of a KL computed against itself: e.g. the snapshot being re-taken after the policy step, the same tensor compared to itself, or a detach/clone that aliases rather than copies.

If that is what is happening, the aux phase is a no-op and PPG is silently degenerating to PPO — which no test would notice, since `ppg_aux_phase_actually_runs` asserts the phase *executes*, not that it *changes* anything.

## Not yet investigated

I did not dig in — it was outside #167's scope and, being identical at HEAD, could not have been caused by it. Flagging it before it is forgotten, per `docs/rules.md` §12.

## Suggested first steps

- Check where the aux-phase target/snapshot distribution is captured relative to the policy-phase optimizer steps in `ppg_agent.rs` (`maybe_aux_phase`).
- Confirm whether the snapshot is a true pre-step copy or a live reference to the stepped policy.
- If the aux phase is confirmed a no-op, the fix likely needs a regression test asserting KL > 0 after a policy step — the current "phase ran" assertion is too weak to protect it.

## Related

- #167 — panic-safety migration during which this surfaced

**Comment by anthonytorlucci** [comment]
## Triage 2026-07-17 — closing as refuted. I filed this; it was a false alarm.

I opened this on the strength of an instrumented run showing the aux-phase KL at bit-exactly `0x00000000` across all 24 minibatches, and argued that an exact zero every minibatch "is not floating-point noise — it reads as a real defect". The reasoning was wrong: I never traced the `24` back to the config that produced it.

**The KL is not computed against itself. PPG is not silently degenerating to PPO.** Printing the KL for *every* aux phase rather than only the last:

```
=== total=2048, n_iteration=4, anneal_lr=true (the configuration I instrumented) ===
aux phase @ iteration  4  lr=1.875000e-4  epochs=6 mbs=24  policy_kl=7.5875944e-4  bits=0x3a46e77c
aux phase @ iteration  8  lr=1.250000e-4  epochs=6 mbs=24  policy_kl=1.4766092e-4  bits=0x391ad56d
aux phase @ iteration 12  lr=6.250000e-5  epochs=6 mbs=24  policy_kl=2.8775134e-5  bits=0x37f16221
aux phase @ iteration 16  lr=0.000000e0   epochs=6 mbs=24  policy_kl=0e0           bits=0x00000000  <-- the zero I reported
=== anneal_lr=false ===
iterations 4/8/12/16 → policy_kl = 1.0290509e-3 / 6.762045e-4 / 4.1657e-4 / 8.639917e-4  (all nonzero)
```

**Why the zero is correct.** `aux_batch_size=128`, `n_iteration=4`, `num_steps=128` → 512 buffered steps → 4 minibatches/epoch × `e_aux=6` epochs = **exactly 24**. So "24 minibatches" was one complete aux phase, read via `last_aux_phase()` — i.e. the iteration-16 phase. With `total=2048 / batch_size=128` → `total_iterations=16`, `annealed_learning_rate` (`ppo_config.rs:347-353`) computes `frac = 1.0 - 16/16 = 0.0` exactly. Adam at lr=0 leaves parameters bit-identical, so `new_logits` are bitwise equal to the `old_logits` snapshot and `Σ p_old·(log p_old − log p_new)` is exactly `0x00000000`. KL's non-negativity means a mean of exactly zero implies every term is exactly zero — fully consistent.

**The snapshot path is correct.** `ppg_agent.rs:655` takes `compute_old_logits_flat` once, at the top of `maybe_aux_phase`, against the current post-policy-phase policy, flattened to a host-side `Vec<f32>` — a true value copy that cannot alias the live module. This matches the reference PPG (`compute_presleep_outputs` before the aux epochs), not a per-minibatch re-snapshot. The trunk is genuinely shared between `logits_head` and `aux_value_head` (`policies/categorical.rs:86-101`), so aux-value gradients do move the logits — which is exactly why the KL grows within a phase and why the nonzero numbers above are what a healthy aux phase looks like.

**Two real things came out of this:**

1. **Aux-phase LR off-by-one → filed as #324.** `maybe_aux_phase` reads `current_learning_rate()` (`ppg_agent.rs:646`) *after* `policy_phase_update` has already done `self.iteration += 1` (`ppg_agent.rs:612`), so the aux phase runs one iteration ahead of the policy phase on the annealing schedule and the final aux phase always lands on exactly 0.0 — 24 minibatches of forward/backward plus its main-value updates, entirely wasted. The policy phase is fine (it snapshots `lr` before the increment, matching CleanRL). CleanRL's schedule never reaches exactly zero; this one does, for the aux phase only.
2. **The regression test is too weak** — my original critique of it stands even though the alarm didn't. `ppg_aux_phase_actually_runs` asserts `aux.policy_kl.is_finite()`, and `0.0` is finite, so it passes on an identically-zero KL. Note the trap for whoever strengthens it: asserting `policy_kl > 0.0` against `last_aux_phase()` under the default config would **fail for the benign reason above**. Assert on a non-final phase, or with `anneal_lr(false)`. Folded into #320 (which already covers the invariant-pinning gap for this crate's tests).

Closing as not planned.

**Comment by anthonytorlucci** [comment]
## The closing comment's forward-looking advice is now refuted by #324

**Leading with the contradiction, because a reviewer following this comment today would write the wrong test.** My 2026-07-17 triage ended with:

> "Note the trap for whoever strengthens it: asserting `policy_kl > 0.0` against `last_aux_phase()` under the default config would **fail for the benign reason above**. Assert on a non-final phase, or with `anneal_lr(false)`."

That advice no longer holds. #324 — the real defect this false alarm surfaced — is fixed on its branch, and the fix removes exactly the "benign reason" the warning rests on.

**What changed.** `maybe_aux_phase` read `current_learning_rate()` after `policy_phase_update` had already done `self.iteration += 1`, so every auxiliary phase annealed one tick ahead of the policy phase it accompanies; at the terminal iteration that tick is exactly `0.0`. The rate the policy phase applies is now snapshotted before the increment and reused by the auxiliary phase:

- `crates/rlevo-reinforcement-learning/src/algorithms/ppg/ppg_agent.rs:582-585` — `let lr = self.current_learning_rate(); self.policy_phase_lr = lr;`, read before the increment.
- `crates/rlevo-reinforcement-learning/src/algorithms/ppg/ppg_agent.rs:785-786` — `let lr = self.policy_phase_lr;` replaces the post-increment read.

**Executed evidence.** The #324 branch ships precisely the test this comment warned would fail — `policy_kl > 0.0` asserted against `last_aux_phase()` under an annealing config where `total_iterations % n_iteration == 0`, i.e. the terminal-phase case (`crates/rlevo/tests/ppg_integration.rs:243-302`):

```
$ cargo test -p rlevo --test ppg_integration ppg_final_aux_phase_steps_at_a_nonzero_learning_rate -- --exact
test ppg_final_aux_phase_steps_at_a_nonzero_learning_rate ... ok
```

It fails on pre-fix code for the right reason — `the final aux phase moved the policy by exactly nothing (policy_kl = 0), which means it ran at lr = 0.0` — verified by reverting only `ppg_agent.rs:786`.

**Why this matters beyond the one test.** Post-fix, a zero `policy_kl` at the final auxiliary phase is a **regression signal, not a benign floor**. Following this comment's advice today would (a) produce a strictly weaker test than the code now supports, and (b) teach a future reader to accept a reintroduction of #324's bug as expected behaviour.

**What still stands from the original triage.** The refutation itself was correct: the KL is not computed against itself, the `compute_old_logits_flat` snapshot path is a true host-side value copy, and PPG was never silently degenerating to PPO. The "24 minibatches" arithmetic is also unaffected — `aux_batch_size=128`, `n_iteration=4`, `num_steps=128` → 512 buffered steps → 4 minibatches/epoch x `e_aux=6` = 24, and #324's diff does not touch that chunking loop, only the rate it steps at.

One correction to #324's own framing while I am here: the one-tick offset is unconditional, but the final auxiliary phase is a *guaranteed* no-op only when `total_iterations % n_iteration == 0`. Otherwise the last auxiliary phase fires before the terminal tick and merely runs at a slightly-too-cold rate.

No state or label change — the closure was and remains correct. This is a record for anyone who reaches the closing comment from #320 or from a search.
